compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -482,74 +491,90 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
 
         let tcx = ecx.tcx();
 
-        let a_ty = goal.predicate.self_ty();
-        let b_ty = goal.predicate.trait_ref.args.type_at(1);
-        let ty::Dynamic(a_data, a_region, ty::Dyn) = *a_ty.kind() else {
-            return vec![];
-        };
-        let ty::Dynamic(b_data, b_region, ty::Dyn) = *b_ty.kind() else {
-            return vec![];
-        };
+        // Need to wrap in a probe since `normalize_non_self_ty` has side-effects.
+        ecx.probe(|_| CandidateKind::DynUpcastingAssembly).enter(|ecx| {
+            let a_ty = goal.predicate.self_ty();
+            let b_ty = goal.predicate.trait_ref.args.type_at(1);
+            let ty::Dynamic(a_data, a_region, ty::Dyn) = *a_ty.kind() else {
+                return vec![];
+            };
 
-        // All of a's auto traits need to be in b's auto traits.
-        let auto_traits_compatible =
-            b_data.auto_traits().all(|b| a_data.auto_traits().any(|a| a == b));
-        if !auto_traits_compatible {
-            return vec![];
-        }
+            // We don't care about `ty::Infer` here or errors here, since we'll
+            // register an ambiguous/error response in the other unsize candidate
+            // assembly function.
```

Comment:
I have to admit that I a somewhat uncomfortable with this and would prefer a match on `b_ty` which also handles `ty::Infer` :sweat_smile:

Reply:
That alternative is better imo, then we'd just ICE in both cases if we encounter an infer var. I'll do that.